### PR TITLE
Add LongCat-Video T2V pipeline to multimodal_gen

### PR DESCRIPTION
**Pull request by Fatemanx**
## Summary

Add text-to-video generation support for [meituan-longcat/LongCat-Video](https://huggingface.co/meituan-longcat/LongCat-Video) to the `multimodal_gen` subsystem.

This PR contributes to #23035

## Components

| File | Description |
|------|-------------|
| `runtime/models/dits/longcat_video.py` | 48-layer DiT with 3D RoPE, FlashAttn2/SDPA attention |
| `runtime/pipelines/longcat_video_pipeline.py` | Pipeline with synthesized model_index (non-standard checkpoint layout) |
| `runtime/pipelines_core/stages/model_specific_stages/longcat_video.py` | Denoising stage with optimized CFG helper |
| `configs/pipeline_configs/longcat_video.py` | BF16 DiT / FP32 VAE, FlowMatch scheduler, linear sigmas |
| `configs/sample/longcat_video.py` | Defaults: 93 frames, 832×480, 50 steps, guidance 4.0 |
| `configs/models/dits/longcat_video.py` | DiT architecture config |

## Design Notes

**Non-standard checkpoint layout**: LongCat-Video uses `dit/` instead of `transformer/` and ships a minimal `model_index.json`. `synthesize_longcat_model_index()` builds a diffusers-compatible index at load time to avoid requiring users to restructure the checkpoint.

**Avatar exclusion**: `meituan-longcat/LongCat-Video-Avatar` is a different model family. Registry detection explicitly excludes paths containing `avatar`.

**Single-GPU MVP**: Multi-GPU SP raises `NotImplementedError`. Frame-level parallelism is a natural extension but out of scope for this PR.

## Supported Resolutions

- 832×480 (landscape)
- 480×832 (portrait)

## Tests

18 unit tests covering sampling params, pipeline config, text postprocess, CFG helper, model index synthesis, registry detection, and error paths.

E2E verified on NVIDIA H20 (96GB):
- 25 frames, 832×480, 10 steps: ~75s, peak 35GB VRAM
- 93 frames, 832×480, 50 steps: supported (default config)

## Bug Fixes (found during E2E)

- **`prepare_latent_shape` double-compression**: `LatentPreparationStage.adjust_video_length` already converts pixel frames → latent frames; `prepare_latent_shape` was re-applying `(n-1)//4+1`, producing 5 output frames instead of 25 for a 25-frame request.

## Checklist

- [x] Unit tests pass (`pytest python/sglang/multimodal_gen/test/unit/test_longcat_video_t2v.py`)
- [x] E2E verified (25-frame video generated correctly)
- [x] Registry integration with Avatar exclusion
- [x] FA2 fallback logs a warning instead of silently degrading
- [x] Input validation for `(num_frames-1)%4==0` constraint

---
### CI States

Latest PR Test (Base): :x: [Run #27220451249](https://github.com/sgl-project/sglang/actions/runs/27220451249)
Latest PR Test (Extra): :x: [Run #27220450620](https://github.com/sgl-project/sglang/actions/runs/27220450620)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/dits/longcat_video.py [hunk 1]
```diff
@@ -0,0 +1,948 @@
...
+        q_, k_ = q.float(), k.float()
+        freqs_cis = freqs_cis.float()
+        cos, sin = freqs_cis.cos(), freqs_cis.sin()
+        cos, sin = rearrange(cos, "n d -> 1 1 n d"), rearrange(sin, "n d -> 1 1 n d")
+        q_ = (q_ * cos) + (rotate_half(q_) * sin)
+        k_ = (k_ * cos) + (rotate_half(k_) * sin)
+        return q_.type_as(q), k_.type_as(k)
+
+
+class Attention(nn.Module):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/dits/longcat_video.py:274]: please adapt to `USPAttention`

File: python/sglang/multimodal_gen/runtime/pipelines/longcat_video.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from sglang.multimodal_gen.runtime.pipelines.longcat_video_pipeline import (
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines/longcat_video.py]: we don't need this necessarily?

File: python/sglang/multimodal_gen/test/e2e/verify_longcat_video_t2v.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
+#!/usr/bin/env python3
```
Comment (mickqian) [inline python/sglang/multimodal_gen/test/e2e/verify_longcat_video_t2v.py]: for e2e test, register your testcase in `gpu_cases.py`

File: python/sglang/multimodal_gen/runtime/models/dits/longcat_video.py (hunk 1 above)
Comment (Fatemanx) [reply python/sglang/multimodal_gen/runtime/models/dits/longcat_video.py:274]: Updated.

  I refactored LongCat self-attention to use `USPAttention` as the main execution path, while keeping the LongCat-specific semantics intact:
  - 3D RoPE
  - q/k norm
  - `num_cond_latents` handling

  I also fixed the LongCat denoising stage so both CFG and non-CFG paths set `forward_context`, which is required by the `USPAttention` path.

  I kept the current LongCat cross-attention implementation unchanged in this PR, since it still uses the model-specific varlen KV flow

File: python/sglang/multimodal_gen/test/e2e/verify_longcat_video_t2v.py (hunk 3 above)
Comment (Fatemanx) [reply python/sglang/multimodal_gen/test/e2e/verify_longcat_video_t2v.py]: Done.

  I registered a LongCat T2V case in `gpu_cases.py` and added the default test model constant in `test_utils.py`.

  The case uses the minimal runnable config for this MVP path:
  - `832x480`
  - `25` frames
  - `10` inference steps
  - `guidance_scale=4.0`

  I also removed the standalone `verify_longcat_video_t2v.py` script so the validation path now goes through the official test harness instead of a
  separate ad hoc entrypoint.

File: python/sglang/multimodal_gen/runtime/pipelines/longcat_video.py (hunk 2 above)
Comment (Fatemanx) [reply python/sglang/multimodal_gen/runtime/pipelines/longcat_video.py]: Removed.

  This wrapper was only re-exporting symbols from `longcat_video_pipeline.py`, and the internal
  registry/tests already use the actual implementation module directly.

File: python/sglang/multimodal_gen/runtime/models/dits/longcat_video.py [hunk 4]
```diff
@@ -0,0 +1,923 @@
...
+
+
+def rotate_half(x: torch.Tensor) -> torch.Tensor:
+    x = rearrange(x, "... (d r) -> ... d r", r=2)
+    x1, x2 = x.unbind(dim=-1)
+    x = torch.stack((-x2, x1), dim=-1)
+    return rearrange(x, "... d r -> ... (d r)")
+
+
+class RotaryPositionalEmbedding(nn.Module):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/dits/longcat_video.py:204]: is it different from existing rope structure?

Comment (Fatemanx) [reply python/sglang/multimodal_gen/runtime/models/dits/longcat_video.py:204]: Thanks. I agree this was duplicating the shared RoPE implementation.

  I updated LongCatVideo to reuse `NDRotaryEmbedding` and the shared `_apply_rotary_emb`, and removed the local RoPE helper
  implementations. The only local part left is grid-position materialization, because
  `NDRotaryEmbedding.forward_from_grid()` currently assumes initialized SP groups, while this LongCat path is no-SP.

  I also added a unit test that checks numerical equivalence with the previous LongCat RoPE math and dtype preservation.
  Validation passed with the LongCat unit tests and an official LongCat T2V 50-step e2e generation run.
